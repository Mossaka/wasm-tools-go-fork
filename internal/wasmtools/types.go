package wasmtools

import (
	"context"
	"io"
	"io/fs"
)

// Runner is an interface for running Wasm modules.
type Runner interface {
	Run(ctx context.Context, args []string, stdin io.Reader, fsMap map[fs.FS]string, name *string) (stdout io.Reader, stderr io.Reader, err error)
}
