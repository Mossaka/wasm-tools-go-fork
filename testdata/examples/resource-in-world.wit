package examples:resources-in-world

interface types {
    resource foo {
        foo: func();
        bar: func(arg: u32);
    }
}

world imports {
    import types;
}
