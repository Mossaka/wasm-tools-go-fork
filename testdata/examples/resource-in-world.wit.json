{
  "worlds": [
    {
      "name": "imports",
      "imports": {
        "interface-0": {
          "interface": 0
        }
      },
      "exports": {},
      "package": 0
    }
  ],
  "interfaces": [
    {
      "name": "types",
      "types": {
        "foo": 0
      },
      "functions": {
        "[method]foo.foo": {
          "name": "[method]foo.foo",
          "kind": {
            "method": 0
          },
          "params": [
            {
              "name": "self",
              "type": 1
            }
          ],
          "results": []
        },
        "[method]foo.bar": {
          "name": "[method]foo.bar",
          "kind": {
            "method": 0
          },
          "params": [
            {
              "name": "self",
              "type": 1
            },
            {
              "name": "arg",
              "type": "u32"
            }
          ],
          "results": []
        }
      },
      "package": 0
    }
  ],
  "types": [
    {
      "name": "foo",
      "kind": "resource",
      "owner": {
        "interface": 0
      }
    },
    {
      "name": null,
      "kind": {
        "handle": {
          "borrow": 0
        }
      },
      "owner": null
    }
  ],
  "packages": [
    {
      "name": "examples:resources-in-world",
      "interfaces": {
        "types": 0
      },
      "worlds": {
        "imports": 0
      }
    }
  ]
}