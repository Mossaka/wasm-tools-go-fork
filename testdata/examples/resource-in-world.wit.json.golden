&wit.Resolve{
    Worlds: {
        &wit.World{
            Name:    "imports",
            Imports: {
                "interface-0": &wit.Interface{
                    Name:     &"types",
                    TypeDefs: {
                        "foo": &wit.TypeDef{
                            Name:       &"foo",
                            Kind:       &wit.Resource{},
                            Owner:      &wit.Interface{(CYCLIC REFERENCE)},
                            Docs:       wit.Docs{},
                            _worldItem: wit._worldItem{},
                            _type:      wit._type{},
                        },
                    },
                    Functions: {
                        "[method]foo.bar": &wit.Function{
                            Name: "[method]foo.bar",
                            Kind: &wit.Method{
                                Type:          &wit.TypeDef{(CYCLIC REFERENCE)},
                                _functionKind: wit._functionKind{},
                            },
                            Params: {
                                {
                                    Name: "self",
                                    Type: &wit.TypeDef{
                                        Name: (*string)(nil),
                                        Kind: &wit.BorrowedHandle{
                                            Type:    &wit.TypeDef{(CYCLIC REFERENCE)},
                                            _handle: wit._handle{},
                                        },
                                        Owner:      nil,
                                        Docs:       wit.Docs{},
                                        _worldItem: wit._worldItem{},
                                        _type:      wit._type{},
                                    },
                                },
                                {
                                    Name: "arg",
                                    Type: wit.U32{},
                                },
                            },
                            Results:    nil,
                            Docs:       wit.Docs{},
                            _worldItem: wit._worldItem{},
                        },
                        "[method]foo.foo": &wit.Function{
                            Name: "[method]foo.foo",
                            Kind: &wit.Method{
                                Type:          &wit.TypeDef{(CYCLIC REFERENCE)},
                                _functionKind: wit._functionKind{},
                            },
                            Params: {
                                {
                                    Name: "self",
                                    Type: &wit.TypeDef{
                                        Name: (*string)(nil),
                                        Kind: &wit.BorrowedHandle{
                                            Type:    &wit.TypeDef{(CYCLIC REFERENCE)},
                                            _handle: wit._handle{},
                                        },
                                        Owner:      nil,
                                        Docs:       wit.Docs{},
                                        _worldItem: wit._worldItem{},
                                        _type:      wit._type{},
                                    },
                                },
                            },
                            Results:    nil,
                            Docs:       wit.Docs{},
                            _worldItem: wit._worldItem{},
                        },
                    },
                    Package: &wit.Package{
                        Name: wit.Ident{
                            Namespace: "examples",
                            Package:   "resources-in-world",
                            Extension: "",
                            Version:   (*semver.Version)(nil),
                            _node:     wit._node{},
                        },
                        Interfaces: {
                            "types": &wit.Interface{(CYCLIC REFERENCE)},
                        },
                        Worlds: {
                            "imports": &wit.World{(CYCLIC REFERENCE)},
                        },
                        Docs: wit.Docs{},
                    },
                    Docs:       wit.Docs{},
                    _typeOwner: wit._typeOwner{},
                    _worldItem: wit._worldItem{},
                },
            },
            Exports: {},
            Package: &wit.Package{
                Name: wit.Ident{
                    Namespace: "examples",
                    Package:   "resources-in-world",
                    Extension: "",
                    Version:   (*semver.Version)(nil),
                    _node:     wit._node{},
                },
                Interfaces: {
                    "types": &wit.Interface{
                        Name:     &"types",
                        TypeDefs: {
                            "foo": &wit.TypeDef{
                                Name:       &"foo",
                                Kind:       &wit.Resource{},
                                Owner:      &wit.Interface{(CYCLIC REFERENCE)},
                                Docs:       wit.Docs{},
                                _worldItem: wit._worldItem{},
                                _type:      wit._type{},
                            },
                        },
                        Functions: {
                            "[method]foo.bar": &wit.Function{
                                Name: "[method]foo.bar",
                                Kind: &wit.Method{
                                    Type:          &wit.TypeDef{(CYCLIC REFERENCE)},
                                    _functionKind: wit._functionKind{},
                                },
                                Params: {
                                    {
                                        Name: "self",
                                        Type: &wit.TypeDef{
                                            Name: (*string)(nil),
                                            Kind: &wit.BorrowedHandle{
                                                Type:    &wit.TypeDef{(CYCLIC REFERENCE)},
                                                _handle: wit._handle{},
                                            },
                                            Owner:      nil,
                                            Docs:       wit.Docs{},
                                            _worldItem: wit._worldItem{},
                                            _type:      wit._type{},
                                        },
                                    },
                                    {
                                        Name: "arg",
                                        Type: wit.U32{},
                                    },
                                },
                                Results:    nil,
                                Docs:       wit.Docs{},
                                _worldItem: wit._worldItem{},
                            },
                            "[method]foo.foo": &wit.Function{
                                Name: "[method]foo.foo",
                                Kind: &wit.Method{
                                    Type:          &wit.TypeDef{(CYCLIC REFERENCE)},
                                    _functionKind: wit._functionKind{},
                                },
                                Params: {
                                    {
                                        Name: "self",
                                        Type: &wit.TypeDef{
                                            Name: (*string)(nil),
                                            Kind: &wit.BorrowedHandle{
                                                Type:    &wit.TypeDef{(CYCLIC REFERENCE)},
                                                _handle: wit._handle{},
                                            },
                                            Owner:      nil,
                                            Docs:       wit.Docs{},
                                            _worldItem: wit._worldItem{},
                                            _type:      wit._type{},
                                        },
                                    },
                                },
                                Results:    nil,
                                Docs:       wit.Docs{},
                                _worldItem: wit._worldItem{},
                            },
                        },
                        Package:    &wit.Package{(CYCLIC REFERENCE)},
                        Docs:       wit.Docs{},
                        _typeOwner: wit._typeOwner{},
                        _worldItem: wit._worldItem{},
                    },
                },
                Worlds: {
                    "imports": &wit.World{(CYCLIC REFERENCE)},
                },
                Docs: wit.Docs{},
            },
            Docs:       wit.Docs{},
            _typeOwner: wit._typeOwner{},
        },
    },
    Interfaces: {
        &wit.Interface{
            Name:     &"types",
            TypeDefs: {
                "foo": &wit.TypeDef{
                    Name:       &"foo",
                    Kind:       &wit.Resource{},
                    Owner:      &wit.Interface{(CYCLIC REFERENCE)},
                    Docs:       wit.Docs{},
                    _worldItem: wit._worldItem{},
                    _type:      wit._type{},
                },
            },
            Functions: {
                "[method]foo.bar": &wit.Function{
                    Name: "[method]foo.bar",
                    Kind: &wit.Method{
                        Type:          &wit.TypeDef{(CYCLIC REFERENCE)},
                        _functionKind: wit._functionKind{},
                    },
                    Params: {
                        {
                            Name: "self",
                            Type: &wit.TypeDef{
                                Name: (*string)(nil),
                                Kind: &wit.BorrowedHandle{
                                    Type:    &wit.TypeDef{(CYCLIC REFERENCE)},
                                    _handle: wit._handle{},
                                },
                                Owner:      nil,
                                Docs:       wit.Docs{},
                                _worldItem: wit._worldItem{},
                                _type:      wit._type{},
                            },
                        },
                        {
                            Name: "arg",
                            Type: wit.U32{},
                        },
                    },
                    Results:    nil,
                    Docs:       wit.Docs{},
                    _worldItem: wit._worldItem{},
                },
                "[method]foo.foo": &wit.Function{
                    Name: "[method]foo.foo",
                    Kind: &wit.Method{
                        Type:          &wit.TypeDef{(CYCLIC REFERENCE)},
                        _functionKind: wit._functionKind{},
                    },
                    Params: {
                        {
                            Name: "self",
                            Type: &wit.TypeDef{
                                Name: (*string)(nil),
                                Kind: &wit.BorrowedHandle{
                                    Type:    &wit.TypeDef{(CYCLIC REFERENCE)},
                                    _handle: wit._handle{},
                                },
                                Owner:      nil,
                                Docs:       wit.Docs{},
                                _worldItem: wit._worldItem{},
                                _type:      wit._type{},
                            },
                        },
                    },
                    Results:    nil,
                    Docs:       wit.Docs{},
                    _worldItem: wit._worldItem{},
                },
            },
            Package: &wit.Package{
                Name: wit.Ident{
                    Namespace: "examples",
                    Package:   "resources-in-world",
                    Extension: "",
                    Version:   (*semver.Version)(nil),
                    _node:     wit._node{},
                },
                Interfaces: {
                    "types": &wit.Interface{(CYCLIC REFERENCE)},
                },
                Worlds: {
                    "imports": &wit.World{(CYCLIC REFERENCE)},
                },
                Docs: wit.Docs{},
            },
            Docs:       wit.Docs{},
            _typeOwner: wit._typeOwner{},
            _worldItem: wit._worldItem{},
        },
    },
    TypeDefs: {
        &wit.TypeDef{
            Name:       &"foo",
            Kind:       &wit.Resource{},
            Owner:      &wit.Interface{(CYCLIC REFERENCE)},
            Docs:       wit.Docs{},
            _worldItem: wit._worldItem{},
            _type:      wit._type{},
        },
        &wit.TypeDef{
            Name: (*string)(nil),
            Kind: &wit.BorrowedHandle{
                Type:    &wit.TypeDef{(CYCLIC REFERENCE)},
                _handle: wit._handle{},
            },
            Owner:      nil,
            Docs:       wit.Docs{},
            _worldItem: wit._worldItem{},
            _type:      wit._type{},
        },
    },
    Packages: {
        &wit.Package{
            Name: wit.Ident{
                Namespace: "examples",
                Package:   "resources-in-world",
                Extension: "",
                Version:   (*semver.Version)(nil),
                _node:     wit._node{},
            },
            Interfaces: {
                "types": &wit.Interface{(CYCLIC REFERENCE)},
            },
            Worlds: {
                "imports": &wit.World{(CYCLIC REFERENCE)},
            },
            Docs: wit.Docs{},
        },
    },
}